src/js_subscriptions.hpp

```diff
@@ -0,0 +1,815 @@
+{
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+    switch (subs->state()) {
+        case sync::SubscriptionSet::State::Uncommitted:
+        case sync::SubscriptionSet::State::Pending:
+        case sync::SubscriptionSet::State::Bootstrapping:
+            return_value.set("pending");
+            break;
+        case sync::SubscriptionSet::State::Complete:
+            return_value.set("complete");
+            break;
+        case sync::SubscriptionSet::State::Error:
+            return_value.set("error");
+            break;
+        case sync::SubscriptionSet::State::Superceded:
+            return_value.set("superceded");
+            break;
+        default:
+            throw std::runtime_error("Unknown state in get_state()");
+    }
+}
+
+/**
+ * @brief Get the version of the subscription set
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param return_value \ref ReturnValue wrapping a number representing the version
+ */
+template <typename T>
+void SubscriptionsClass<T>::get_version(ContextType ctx, ObjectType this_object, ReturnValue& return_value)
+{
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+    return_value.set((uint32_t)subs->version());
+}
+
+/**
+ * @brief Get an array snapshot of the subscriptions in the subscription set
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param args \ref None
+ * @param return_value \ref ReturnValue wrapping an array of Subscription instances
+ */
+template <typename T>
+void SubscriptionsClass<T>::snapshot(ContextType ctx, ObjectType this_object, Arguments& args,
+                                     ReturnValue& return_value)
+{
+    args.validate_count(0);
+
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+
+    auto js_subs = std::vector<ValueType>();
+    for (auto& sub : *subs) {
+        js_subs.emplace_back(SubscriptionClass<T>::create_instance(ctx, sub));
+    }
+
+    auto subs_array = Object::create_array(ctx, js_subs);
+    return_value.set(subs_array);
+}
+
+/**
+ * @brief Find a subscription by name
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param args \ref A single argument containing the subscription name to find
+ * @param return_value \ref ReturnValue wrapping a Subscription if found, null if not
+ */
+template <typename T>
+void SubscriptionsClass<T>::find_by_name(ContextType ctx, ObjectType this_object, Arguments& args,
+                                         ReturnValue& return_value)
+{
+    args.validate_count(1);
+
+    std::string name = Value::validated_to_string(ctx, args[0], "name");
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+
+    auto it = subs->find(name);
+
+    if (it != subs->end()) {
```

Comment:
Is the negated version easier to understand:

```
if (it == subs->end()) {
  return_value.set_null();
}
else {
  return_value.set(sub);
}
```

Reply:
I guess you could argue it either way, I have no strong feelings 😁  any preferences @kraenhansen @takameyer ?

Reply:
I vote for the negated version 👍🏼

Reply:
I agree - one less not (`!`) to think about.